Source/EasyNetQ/Consumer/DefaultConsumerErrorStrategy.cs

```diff
@@ -200,12 +210,20 @@ private string CreateConnectionCheckMessage()
         }
 
         private bool disposed = false;
+        private bool disposing = false;
 
         public virtual void Dispose()
         {
             if (disposed) return;
+            disposing = true;
 
-            if(connection != null) connection.Dispose();
+            lock (syncLock)
```

Comment:
@micdenny true, lock in Dispose is probably bad: it could block the finalizer thread.  

It was added to avoid a complex (observed in debugging) race condition between `Dispose()` and `Connect()` where `if(connection != null) connection.Dispose();` did nothing (because connection was still null), but the `Connect()` method had already passed the new `&& !(disposing || disposed)` condition, and so continued to make a new `connection = connectionFactory.CreateConnection();` (this connection never got disposed and still held the application open).  The lock avoided this, but I agree lock in Dispose seems bad.  This particular race condition can probably be fixed another way.  I will propose an alternative shortly, which removes the lock from Dispose() method.